hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/InsertIntoHoodieTableCommand.scala

```diff
@@ -209,19 +209,32 @@ object InsertIntoHoodieTableCommand {
       .getOrElse(INSERT_DROP_DUPS_OPT_KEY.defaultValue)
       .toBoolean
 
-    val operation = if (isOverwrite) {
-      if (table.partitionColumnNames.nonEmpty) {
-        INSERT_OVERWRITE_OPERATION_OPT_VAL  // overwrite partition
-      } else {
-        INSERT_OPERATION_OPT_VAL
+    val enableBulkInsert = parameters.getOrElse(DataSourceWriteOptions.SQL_ENABLE_BULK_INSERT.key,
+      DataSourceWriteOptions.SQL_ENABLE_BULK_INSERT.defaultValue()).toBoolean
+    val isPartitionedTable = table.partitionColumnNames.nonEmpty
+    val isPrimaryKeyTable = primaryColumns.nonEmpty
+    val operation =
+      (isPrimaryKeyTable, enableBulkInsert, isOverwrite, dropDuplicate) match {
+        case (true, true, _, _) =>
+          throw new IllegalArgumentException(s"Table with primaryKey can not use bulk insert.")
```

Comment:
Because currently we will do the primary key uniqueness check when inserting data to the pk-table, just like what database does. Bulk insert currently cannot do such things.

Reply:
hmmm, interesting. I didn't know we do uniqueness check for inserts with primary keyed table.

Reply:
anyways, we can all it out that its responsibility of the user to ensure there are uniqueness. Also, IIUC, hudi can handle duplicates. Incase of updates, both records will be updated. but bulk_insert is very performant compared to regular Insert especially w/ row wirter. So, we should not keep it too restrictive for use. I know from the community msgs, that lot of users leverage bulk_insert. I would vote to relax this constraint.

Reply:
+1 users might have another hudi table for e.g to CTAS from. So if we disallow bulk insert with a pk, then there is no good way to do a full bootstrap. Can we relax this?